src/java.ts

```diff
@@ -0,0 +1,160 @@
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+import fs from 'fs';
+import path from 'path';
+import semver, { SemVer } from 'semver';
+import {
+  API_OLD_VERSION_ENDPOINT,
+  API_V2_JRE_ENDPOINT,
+  API_V2_VERSION_ENDPOINT,
+  SONAR_CACHE_DIR,
+  SONARQUBE_JRE_PROVISIONING_MIN_VERSION,
+  UNARCHIVE_SUFFIX,
+} from './constants';
+import { extractArchive, getCachedFileLocation, validateChecksum } from './file';
+import { log, LogLevel } from './logging';
+import { download, fetch } from './request';
+import { JREFullData, PlatformInfo, ScannerProperties, ScannerProperty } from './types';
+
+export async function fetchServerVersion(): Promise<SemVer> {
+  let version: SemVer | null = null;
+  try {
+    // Try and fetch the new version endpoint first
+    log(LogLevel.DEBUG, `Fetching API V2 ${API_V2_VERSION_ENDPOINT}`);
+    const response = await fetch({
+      url: API_V2_VERSION_ENDPOINT,
+    });
+    version = semver.coerce(response.data);
+  } catch (error: unknown) {
+    try {
+      // If it fails, fallback on deprecated server version endpoint
+      log(
+        LogLevel.DEBUG,
+        `Unable to fetch API V2 ${API_V2_VERSION_ENDPOINT}: ${error}. Falling back on ${API_OLD_VERSION_ENDPOINT}`,
+      );
+      const response = await fetch({
+        url: API_OLD_VERSION_ENDPOINT,
+      });
+      version = semver.coerce(response.data);
+    } catch (error: unknown) {
+      // If it also failed, give up
+      log(LogLevel.ERROR, `Failed to fetch server version: ${error}`);
+      throw error;
+    }
+  }
+
+  // If we couldn't parse the version
+  if (!version) {
+    throw new Error(`Failed to parse server version "${version}"`);
+  }
+
+  return version;
+}
+
+export async function serverSupportsJREProvisioning(
+  parameters: ScannerProperties,
+): Promise<boolean> {
+  if (parameters[ScannerProperty.SonarScannerInternalIsSonarCloud] === 'true') {
+    return true;
+  }
+
+  // SonarQube
+  log(LogLevel.DEBUG, 'Detecting SonarQube server version');
+  const SQServerInfo =
+    semver.coerce(parameters[ScannerProperty.SonarScannerInternalSqVersion]) ??
+    (await fetchServerVersion());
+  log(LogLevel.INFO, 'SonarQube server version: ', SQServerInfo.version);
+
+  const supports = semver.satisfies(SQServerInfo, `>=${SONARQUBE_JRE_PROVISIONING_MIN_VERSION}`);
```

Comment:
question here: is it possible for a server to fail on fetchServerVersion with V2, succeed with old version, and be >=10.5?

Reply:
So there are two things here:
- You mention `10.5` because we use `10.5` in the code but this is incorrect, it should be `10.6`, I will update it
- The only possible case I can see is the user provides an invalid token:
  - the V2 endpoint requires authentication and will fail
  - the V1 endpoint (`/api/server/version`) does not require token and will succeed
  - the analysis will fail in a later step because the token is not valid